### PR TITLE
Test branch

### PR DESCRIPTION
**Pull request by DoMa157**

### CODE REVIEW

**Review by memory-hunter — APPROVED** [review]
looking good

File: src/main/java/math/MatrixHelper.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
+package math;
```
Comment (DerWaschbar) [inline src/main/java/math/MatrixHelper.java:1]: Why is this change in test PR?

File: src/main/java/math/Transform.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package math;
```
Comment (DerWaschbar) [inline src/main/java/math/Transform.java:1]: Why is this change in test PR?

File: src/test/java/math/Matrix4x4Test.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+            {2,0,1,0},
+            {3,0,5,0},
+            {0,1,2,2}
+    });
+    Matrix4x4 m2 = new Matrix4x4(new float[][]{
+            {3,1,1,0},
+            {5,0,0,1},
+            {4,0,1,5},
+            {3,1,5,6}
+    });
```
Comment (DerWaschbar) [inline src/test/java/math/Matrix4x4Test.java:19]: Create `@beforeAll` init function and put all initialization there. It's important to make sure that one test does not change input data and affect others.

~~~
@beforeAll
void init() {
  ....
  }
  ~~~

File: src/test/java/math/Matrix4x4Test.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    @Test
+    void multiply(){
```
Comment (DerWaschbar) [inline src/test/java/math/Matrix4x4Test.java:21]:
```suggestion
    void testMultiply(){
```

File: src/test/java/math/Matrix4x4Test.java [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        Matrix4x4 testMultMatrix = new Matrix4x4(new float[][]{
+                {6,2,6,6},
+                {10,2,3,5},
+                {29,3,8,25},
+                {19,2,12,23}
+        });
+        assertEquals(testMultMatrix,m1.multiply(m2));
+    }
+    @Test
+    void transpose(){
```
Comment (DerWaschbar) [inline src/test/java/math/Matrix4x4Test.java:31]:
```suggestion
    void testTranspose(){
```

File: src/test/java/math/Matrix4x4Test.java [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        Matrix4x4 testTransMatrix = new Matrix4x4(new float[][]{
+                {1,2,3,0},
+                {0,0,0,1},
+                {0,1,5,2},
+                {1,0,0,2}
+        });
+        assertEquals(testTransMatrix,m1.transpose());
+    }
+    @Test
+    void setIdentity(){
```
Comment (DerWaschbar) [inline src/test/java/math/Matrix4x4Test.java:41]:
```suggestion
    void testSetIdentity(){
```

File: src/test/java/math/Matrix4x4Test.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
+package math;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+
+public class Matrix4x4Test {
```
Comment (DerWaschbar) [inline src/test/java/math/Matrix4x4Test.java:7]: `Matrix4x4` has other methods too, like - `zero`, `setCell` `negate` and others. Add tests for all of them.

File: src/test/java/math/Vector4fTest.java [hunk 8]
```diff
@@ -0,0 +1,13 @@
+package math;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class Vector4fTest {
+    Vector4f v1 = new Vector4f(2,2,2,2);
+    Vector4f v2 = new Vector4f(3,3,3,3);
```
Comment (DerWaschbar) [inline src/test/java/math/Vector4fTest.java:7]: Should be init function as well.

File: src/test/java/math/Vector4fTest.java [hunk 9]
```diff
@@ -0,0 +1,13 @@
...
+    @Test
+    void normalize(){
```
Comment (DerWaschbar) [inline src/test/java/math/Vector4fTest.java:9]:
```suggestion
    void testNormalize(){
```

File: src/test/java/math/Vector3fTest.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
+package math;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class Vector3fTest {
+    Vector3f v1 = new Vector3f(2,2,2);
+    Vector3f v2 = new Vector3f(7,3,9);
+    Vector3f v3 = new Vector3f(2,2,1);
```
Comment (DerWaschbar) [inline src/test/java/math/Vector3fTest.java:10]: Should be in init method!

File: src/test/java/math/Vector3fTest.java [hunk 11]
```diff
@@ -0,0 +1,27 @@
+package math;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class Vector3fTest {
```
Comment (DerWaschbar) [inline src/test/java/math/Vector3fTest.java:7]: What about other methods? Add tests for more of them.

Test method names should include 'test' in them.

**Review by DerWaschbar — CHANGES_REQUESTED** [review]

File: src/main/java/math/Matrix4x4.java [hunk 12]
```diff
@@ -112,99 +103,106 @@ public Matrix4x4 inverse() throws IllegalArgumentException {
...
+    }
+
+    public boolean equals(Object obj){
+        float epsilon = 0.000001f;
+        if(obj instanceof Matrix4x4){
+            for(int i=0;i<4;i++){
+                for(int j=0;j<4;j++){
+                    if((((Matrix4x4) obj).getCell(i,j)==-0f&&this.getCell(i,j)==0f)||((Matrix4x4) obj).getCell(i,j)==0f&&this.getCell(i,j)==-0f){
+                        //do nothing
+                    }
```
Comment (DerWaschbar) [inline src/main/java/math/Matrix4x4.java:197]: ??????

Comment (DerWaschbar) [reply src/main/java/math/Matrix4x4.java:197]: Clean up if condition's statement. Remove if not necessary or implement the body.

**Review by DerWaschbar — CHANGES_REQUESTED** [review]

File: src/main/java/math/Matrix4x4.java (hunk 12 above)
Comment (DerWaschbar) [reply src/main/java/math/Matrix4x4.java:197]: Fix & Merge, do not wait for my review, teammates can approve it.

**Review by memory-hunter — APPROVED** [review]
Looking good, fixes have been made and tests pass

**Review by DerWaschbar — APPROVED** [review]

**Pull request merged**